core/src/main/java/io/grpc/internal/ClientCallImpl.java

```diff
@@ -214,57 +247,51 @@ public void runInContext() {
     if (compressor != Codec.Identity.NONE) {
       stream.setMessageCompression(true);
     }
+
     // Delay any sources of cancellation after start(), because most of the transports are broken if
     // they receive cancel before start. Issue #1343 has more details
-
-    // Start the deadline timer after stream creation because it will close the stream
-    Long timeoutNanos = getRemainingTimeoutNanos(callOptions.getDeadlineNanoTime());
-    if (timeoutNanos != null) {
-      deadlineCancellationFuture = startDeadlineTimer(timeoutNanos);
-      if (deadlineCancellationFutureShouldBeCancelled) {
-        // Race detected! ClientStreamListener.closed may have been called before
-        // deadlineCancellationFuture was set, thereby preventing the future from being cancelled.
-        // Go ahead and cancel again, just to be sure it was cancelled.
-        deadlineCancellationFuture.cancel(false);
-      }
-    }
     // Propagate later Context cancellation to the remote side.
-    this.context.addListener(this, directExecutor());
+    context.addListener(this, directExecutor());
+    if (contextListenerShouldBeRemoved) {
+      // Race detected! ClientStreamListener.closed may have been called before
+      // deadlineCancellationFuture was set, thereby preventing the future from being cancelled.
+      // Go ahead and cancel again, just to be sure it was cancelled.
+      context.removeListener(this);
+    }
   }
 
   /**
    * Based on the deadline, calculate and set the timeout to the given headers.
-   *
-   * @return {@code false} if deadline already exceeded
    */
-  static boolean updateTimeoutHeader(@Nullable Long deadlineNanoTime, Metadata headers) {
-    // Fill out timeout on the headers
-    // TODO(carl-mastrangelo): Find out if this should always remove the timeout,
-    // even when returning false.
+  private static void updateTimeoutHeaders(final long now, @Nullable Deadline effectiveDeadline,
+      @Nullable Deadline callDeadline, @Nullable Deadline outerCallDeadline, Metadata headers) {
     headers.removeAll(TIMEOUT_KEY);
-    // Convert the deadline to timeout. Timeout is more favorable than deadline on the wire
-    // because timeout tolerates the clock difference between machines.
-    Long timeoutNanos = getRemainingTimeoutNanos(deadlineNanoTime);
-    if (timeoutNanos != null) {
-      if (timeoutNanos <= 0) {
-        return false;
-      }
-      headers.put(TIMEOUT_KEY, timeoutNanos);
+    headers.removeAll(SERVER_CALL_TIMEOUT_KEY);
+    headers.removeAll(CLIENT_CALL_TIMEOUT_KEY);
+
+    if (effectiveDeadline != null) {
+      long effectiveTimeout = effectiveDeadline.deadlineNanos() - now;
+      headers.put(TIMEOUT_KEY, effectiveTimeout);
+    }
+
+    // Include both deadlines/timeouts for debugging purposes.
```

Comment:
Why do we want these propagating on the wire? Wouldn't logging / recording in the created context satisfy the same need?